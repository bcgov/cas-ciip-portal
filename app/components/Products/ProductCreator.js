import React, {Component} from 'react';
import {graphql, createFragmentContainer, commitMutation} from 'react-relay';
import {Form, Button, Col} from 'react-bootstrap';

class ProductCreator extends Component {
  constructor(props) {
    super(props);
    this.createProductFromRef = React.createRef();
    this.createProduct = graphql`
      mutation ProductCreatorMutation($input: CreateProductInput!) {
        createProduct(input: $input) {
          product {
            rowId
          }
        }
      }
    `;
  }

  saveProduct = event => {
    event.preventDefault();
    event.stopPropagation();
    const date = new Date().toUTCString();
    const saveVariables = {
      input: {
        product: {
          name: event.target.product_name.value,
          description: event.target.product_description.value,
          state: 'active',
          parent: [null],
          createdAt: date,
          createdBy: 'Admin'
        }
      }
    };

    const saveMutation = this.createProduct;
    const {environment} = this.props.relay;
    commitMutation(environment, {
      mutation: saveMutation,
      variables: saveVariables,
      onCompleted: response => {
        console.log(response);
        this.createProductFromRef.current.reset();
        window.location.reload();
      },
      onError: err => console.error(err)
    });
  };

  render() {
    return (
      <>
        <div>
          <Form ref={this.createProductFromRef} onSubmit={this.saveProduct}>
            <Form.Row>
              <Form.Group as={Col} md="4" controlId="product_name">
                <Form.Label>Product Name</Form.Label>
                <Form.Control required="required" type="text" placeholder="" />
              </Form.Group>
              <Form.Group as={Col} md="8" controlId="product_description">
                <Form.Label>Product Description</Form.Label>
                <Form.Control
                  required="required"
                  type="textbox"
                  placeholder=""
                />
              </Form.Group>
            </Form.Row>
            <Button type="submit">Create Product</Button>
          </Form>
        </div>
      </>
    );
  }
}

export default createFragmentContainer(ProductCreator, {});
