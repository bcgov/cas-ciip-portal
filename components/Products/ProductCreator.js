import React ,{ Component } from 'react'
import {graphql, QueryRenderer, commitMutation} from "react-relay";
import { Form, Button, Col } from 'react-bootstrap';
import initEnvironment from '../../lib/createRelayEnvironment';
const environment = initEnvironment();

class ProductCreator extends Component {

    constructor(props) {
        super(props);
        this.createProductFromRef = React.createRef();
        this.createProduct = graphql`
            mutation ProductCreatorMutation ($input: CreateProductInput!){
                createProduct(input:$input){
                    product{
                        rowId
                    }
                }
            }
        `;

        this.createProductBenchmark = graphql`
            mutation ProductCreatorProductBenchmarkMutation ($input: CreateProductBenchmarkInput!){
                createProductBenchmark(input:$input){
                    productBenchmark{
                        rowId
                    }
                }
            }
        `;

        this.updateProductBenchmark = graphql`
            mutation ProductCreatorUpdateProductBenchmarkMutation ($input: UpdateProductBenchmarkByRowIdInput!){
                updateProductBenchmarkByRowId(input:$input){
                    productBenchmark{
                        rowId
                    }
                }
            }
        `;
    }

    updateProductBenchmarkProductId = (productBenchmarkId, productId) => {

        const saveVariables =
            {
                "input": {
                    "productBenchmarkPatch": {
                        "productId": productId
                    },
                    "rowId": productBenchmarkId
                },
            }
        const saveMutation = this.updateProductBenchmark;
        commitMutation(
            environment,
            {
                mutation: saveMutation,
                variables: saveVariables,
                onCompleted: (response, errors) => {
                    console.log(response);
                },
                onError: err => console.error(err),
            },
      ) ;
      window.location.reload();
    }

    saveProduct = (product, productBenchmarkId) => {

      const saveVariables =
          {
              "input": {
                  "product": {
                      "productBenchmarkId": productBenchmarkId,
                      "name": product.product_name,
                      "description": product.product_desc,
                      "archived": false
                  }
              }
          };

      const saveMutation = this.createProduct;
      commitMutation(
          environment,
          {
              mutation: saveMutation,
              variables: saveVariables,
              onCompleted: (response, errors) => {
                  console.log(response);
                  this.updateProductBenchmarkProductId(productBenchmarkId, response.createProduct.product.rowId);
                  this.createProductFromRef.current.reset();
              },
              onError: err => console.error(err),
          },
      );
  }

    saveProductBenchmark = (event) => {
        event.preventDefault();
        event.stopPropagation();
        const product_name = event.target.product_name.value;
        const product_desc = event.target.product_description.value;
        const saveVariables =
            {
                "input": {
                    "productBenchmark": {}
                }
            };

        const saveMutation = this.createProductBenchmark;
        commitMutation(
            environment,
            {
                mutation: saveMutation,
                variables: saveVariables,
                onCompleted: (response, errors) => {
                    console.log(response);
                    this.saveProduct({product_name, product_desc}, response.createProductBenchmark.productBenchmark.rowId)
                },
                onError: err => console.error(err),
            },
        );
    }

    render() {
        return (
            <React.Fragment>
                <div>
                    <Form ref={ this.createProductFromRef } onSubmit={this.saveProductBenchmark}>
                        <Form.Row>
                        <Form.Group as={Col} md="4" controlId="product_name">
                            <Form.Label>Product Name</Form.Label>
                            <Form.Control required="required" type="text" placeholder="" />
                        </Form.Group>
                        <Form.Group as={Col} md="8" controlId="product_description">
                            <Form.Label>Product Description</Form.Label>
                            <Form.Control required="required" type="textbox" placeholder="" />
                        </Form.Group>
                        </Form.Row>
                        <Button type="submit">Create Product</Button>
                    </Form>
                </div>
            </React.Fragment>
        );
    }

}

export default ProductCreator;