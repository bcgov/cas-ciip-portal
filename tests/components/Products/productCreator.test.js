import React from 'react';
import ProductCreator from '../../../components/Products/ProductCreator';
import { render } from '@testing-library/react';



describe('Product Creator', () => {

    it('should render the product name field', async () => {
        // This will replace the query in ProductList with the one above and wait till Milk is rendered
        const {getByLabelText} = render(<ProductCreator />);
        expect(getByLabelText(/Product Name/i)).toBeDefined();
    });

    it('should render the product desc field', async () => {
        // This will replace the query in ProductList with the one above and wait till Milk is rendered
        const {getByLabelText} = render(<ProductCreator />);
        expect(getByLabelText(/Product Description/i)).toBeDefined();
    });

    it('should render the Create Product button', async () => {
        // This will replace the query in ProductList with the one above and wait till Milk is rendered
        const {getByText} = render(<ProductCreator />);
        expect(getByText(/Create Product/i)).toBeDefined();
    });

    it('should match the last usable snapshot', async () => {
        // This will replace the query in ProductList with the one above and wait till Milk is rendered
        const container = render(<ProductCreator />);
        expect(container).toMatchSnapshot();
    });

});

// Test the productCreator Mutation