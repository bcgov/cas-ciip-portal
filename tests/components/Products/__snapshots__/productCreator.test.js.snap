// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Creator should match the last usable snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <form
          class=""
        >
          <div
            class="form-row"
          >
            <div
              class="form-group col-md-4"
            >
              <label
                class="form-label"
                for="product_name"
              >
                Product Name
              </label>
              <input
                class="form-control"
                id="product_name"
                placeholder=""
                required=""
                type="text"
              />
            </div>
            <div
              class="form-group col-md-8"
            >
              <label
                class="form-label"
                for="product_description"
              >
                Product Description
              </label>
              <input
                class="form-control"
                id="product_description"
                placeholder=""
                required=""
                type="textbox"
              />
            </div>
          </div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Create Product
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <form
        class=""
      >
        <div
          class="form-row"
        >
          <div
            class="form-group col-md-4"
          >
            <label
              class="form-label"
              for="product_name"
            >
              Product Name
            </label>
            <input
              class="form-control"
              id="product_name"
              placeholder=""
              required=""
              type="text"
            />
          </div>
          <div
            class="form-group col-md-8"
          >
            <label
              class="form-label"
              for="product_description"
            >
              Product Description
            </label>
            <input
              class="form-control"
              id="product_description"
              placeholder=""
              required=""
              type="textbox"
            />
          </div>
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Create Product
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
